### PR TITLE
[schema] Support database path for specified database in catalog

### LINKED ISSUE #2493: [Feature][schema] Support database path for specified database in catalog
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Subtask of #2377 

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by FangYongs**
### Purpose

Linked issue: close #2493 

Currently there will be `FileIO` when a catalog is created. We would like to support path for `database` and paimon need to get fileio for each database when it is created, dropped and tables are created, dropped for it. This PR aims to introduce `DatabaseFileIOProvider` in catalog to get fileio for given database name.

### Tests

Exist tests can cover this feature.

### API and Format

no

### Documentation

no

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Hi @FangYongs 
Does this come from real business needs? In my previous understanding, this situation is usually solved by the FileSystem of HDFS. (Or Flink FileSystem)

**Comment by FangYongs** [comment]
> Hi @FangYongs Does this come from real business needs? In my previous understanding, this situation is usually solved by the FileSystem of HDFS. (Or Flink FileSystem)

@JingsongLi Yes. Currently in our production environment, we always support one catalog for hive, hudi, iceberg or paimon, so they can easily read/write tables use `paimon`.`databaseName`.`tableName`. We have a meta service in company to manage all databases and tables for each catalog. When users create a database, they need to give a path of hdfs/s3 for it and then create tables in the database. So we need to support configuring warehosue path in database level.

**Comment by JingsongLi** [comment]
> > Hi @FangYongs Does this come from real business needs? In my previous understanding, this situation is usually solved by the FileSystem of HDFS. (Or Flink FileSystem)
> 
> @JingsongLi Yes. Currently in our production environment, we always support one catalog for hive, hudi, iceberg or paimon, so they can easily read/write tables use `paimon`.`databaseName`.`tableName`. We have a meta service in company to manage all databases and tables for each catalog. When users create a database, they need to give a path of hdfs/s3 for it and then create tables in the database. So we need to support configuring warehosue path in database level.

Can Hadoop FileSystem solve this? https://github.com/apache/incubator-paimon/pull/2504

**Comment by FangYongs** [comment]
@JingsongLi It seems not,  `HadoopFileIO` is for catalog too. As mentioned above, we have a meta service for all catalogs, so we create a `paimon meta catalog` based on it, and the catalog stores the database and path after users created it. When users create a table, it needs to find the database and its path from the `paimon meta catalog` and create `FileIO` for the table to store schema and data, that's why I would like to introduce the interface `DatabaseFileIOProvider` in `AbstractCatalog` to get `FileIO` for specified database. 

What do you think?

**Comment by JingsongLi** [comment]
> @JingsongLi It seems not, `HadoopFileIO` is for catalog too. As mentioned above, we have a meta service for all catalogs, so we create a `paimon meta catalog` based on it, and the catalog stores the database and path after users created it. When users create a table, it needs to find the database and its path from the `paimon meta catalog` and create `FileIO` for the table to store schema and data, that's why I would like to introduce the interface `DatabaseFileIOProvider` in `AbstractCatalog` to get `FileIO` for specified database.
> 
> What do you think?

Can you list what FileSystems need to be supported?

**Comment by JingsongLi** [comment]
For example, Hadoop FileSystem supports viewfs, there can be multiple underly filesystems.

**Comment by FangYongs** [comment]
@JingsongLi Local file system, hdfs or s3 which have beed supported in paimon

**Comment by FangYongs** [comment]
> For example, Hadoop FileSystem supports viewfs, there can be multiple underly filesystems.

We may configure different hdfs cluster for different databases in the paimon catalog, do you mean we can get different filesystem from one `FileIO`?

**Comment by JingsongLi** [comment]
> > For example, Hadoop FileSystem supports viewfs, there can be multiple underly filesystems.
> 
> We may configure different hdfs cluster for different databases in the paimon catalog, do you mean we can get different filesystem from one `FileIO`?

Yes

**Comment by JingsongLi** [comment]
From Paimon Level, you can do this too. Just like current `HadoopFileIO`, there is a `Map<Pair<String, String>, FileSystem> fsMap`, which means scheme + authority => FileSystem.

**Comment by FangYongs** [comment]
@JingsongLi Get it. Then I think we can add minor updates in paimon to support different database paths for catalog, I will update this PR

**Comment by FangYongs** [comment]
@JingsongLi What do you think of this PR now? We can implement customized catalog and get warehouse path with specified database name

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java [hunk 1]
```diff
@@ -296,7 +296,7 @@ private FileStoreTable getDataTable(Identifier identifier) throws TableNotExistE
 
     @VisibleForTesting
     public Path newDatabasePath(String database) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:298]: You can just override this method?
We can just remove `VisibleForTesting`, and add documentation.

Comment (FangYongs) [reply paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:304]: Sounds reasonable, I have updated this PR

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java [hunk 2]
```diff
@@ -326,7 +337,7 @@ protected abstract TableSchema getDataTableSchema(Identifier identifier)
 
     @VisibleForTesting
     public Path getDataTableLocation(Identifier identifier) {
-        return newTableLocation(warehouse(), identifier);
+        return newTableLocation(newDatabasePath(identifier.getDatabaseName()), identifier);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:340]: I'm a bit confused about how much code we need to modify. Should we just handle it directly here.

Just: `new Path(databasePath, tableName)` here?

**Comment by FangYongs** [comment]
@JingsongLi Can you help to review this PR again? THX

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**